### PR TITLE
Support Anomaly Config in SLO Spec

### PR DESCRIPTION
**Pull request by nuusk**
## Before merging!
- [x] Merge [This](https://github.com/nobl9/nobl9-go/pull/37) nobl9-go PR 
- [x] Release new nobl9-go version
- [x] Update nobl9-go version in go.mod in this repo

## Motivation
This PR adds support for new field `anomaly_config` in SLO Spec that will allow users to define Alert Methods they want to be notified on when their SLO meets the criteria of Anomaly Detection - No Data (Objective not gathering data for more than 15 minutes).

## Local replace
Add this line to `go.mod` in order to work with local instance of `nobl9-go` repository 
```
replace github.com/nobl9/nobl9-go v0.8.1 => ../nobl9-go
```

## Testing
Add `anomaly_config` section to any SLO using terraform.
```
resource "nobl9_slo" "this" {
  name             = "${nobl9_project.this.name}-latency"
  service          = nobl9_service.this.name
  budgeting_method = "Occurrences"
  project          = nobl9_project.this.name

  time_window {
    unit       = "Day"
    count      = 30
    is_rolling = true
  }

  objective {
    name         = "tf-objective-1"
    target       = 0.99
    display_name = "OK"
    value        = 2000
    op           = "gte"

    raw_metric {
      query {
        prometheus {
          promql = <<EOT
          latency_west_c7{code="ALL",instance="localhost:3000",job="prometheus",service="globacount"}
          EOT
        }
      }
    }
  }

  indicator {
    name = "test-terraform-prom-agent"
    project = "test-n9-terraform"
  }

  anomaly_config {
    no_data {
      alert_method {
        name = "slack-notification"
        project = "default"
      }
    }
  }
}
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -10,6 +10,8 @@ require (
 	github.com/nobl9/nobl9-go v0.7.3
 )
 
+replace github.com/nobl9/nobl9-go v0.7.3 => ../nobl9-go
```
Comment (nuusk) [inline go.mod:13]: This will be removed before merging and releasing nobl9-go

File: nobl9/anomaly_config_schema.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+//	  "method3": { "name": "method3", "project": "project1" },
+//	}
+func transformAnomalyConfigAlertMethodsTo2DMap(alertMethods []n9api.AnomalyConfigAlertMethod) map[string]map[string]string {
+	result := make(map[string]map[string]string)
+	for _, method := range alertMethods {
+		values := make(map[string]string)
+
+		values["name"] = method.Name
+		values["project"] = method.Project
+		result[method.Name] = values
```
Comment (piotrkwarcinski) [inline nobl9/anomaly_config_schema.go:187]:
```suggestion
		result[method.Name] = map[string]string{
			"name": method.Name,
			"project": method.Project,
		}
```
For me it's more readable, but you can leave it the way it is, only maybe remove the empty line.

File: nobl9/resource_slo.go [hunk 3]
```diff
@@ -770,7 +776,8 @@ func unmarshalTimeWindow(d *schema.ResourceData, spec map[string]interface{}) er
 		calendarTF["time_zone"] = calendar["timeZone"].(string)
 		timeWindowsTF["calendar"] = schema.NewSet(oneElementSet, []interface{}{calendarTF})
 	}
-	return d.Set("time_window", schema.NewSet(oneElementSet, []interface{}{timeWindowsTF}))
+	y := schema.NewSet(oneElementSet, []interface{}{timeWindowsTF})
+	return d.Set("time_window", y)
```
Comment (piotrkwarcinski) [inline nobl9/resource_slo.go:780]: How about?
```suggestion
	tw := schema.NewSet(oneElementSet, []interface{}{timeWindowsTF})
	return d.Set("time_window", tw)
```

**Review by piotrkwarcinski — CHANGES_REQUESTED** [review]
I left a few suggestions for now :), when it's ready I'll do the review again. I didn't comment on commented tests and  the `TF_ACC` flag but I suspect those are temporary changes and will be fixed when the PR is ready.
Btw. I tried to run the current version of the code and it doesn't seem to work:
```
Stack trace from the terraform-provider-nobl9 plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 71 [running]:
github.com/nobl9/terraform-provider-nobl9/nobl9.unmarshalAnomalyConfig(0xc00020b580?, 0xc8608c?)
        /home/piotr/dev/terraform-provider-nobl9/nobl9/anomaly_config_schema.go:131 +0x45c
github.com/nobl9/terraform-provider-nobl9/nobl9.unmarshalSLO(0xc0006861c8?, {0xc0004f8040?, 0x3?, 0x0?})
        /home/piotr/dev/terraform-provider-nobl9/nobl9/resource_slo.go:702 +0xfad
github.com/nobl9/terraform-provider-nobl9/nobl9.resourceSLORead({0xdbdc60?, 0xc000249830?}, 0xc00020b580, {0xc48940?, 0xc00068e230?})
        /home/piotr/dev/terraform-provider-nobl9/nobl9/resource_slo.go:372 +0x305
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0004abea0, {0xdbdc60, 0xc000249830}, 0xd?, {0xc48940, 0xc00068e230})
        /home/piotr/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.19.0/helper/schema/resource.go:724 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc0004abea0, {0xdbdc60, 0xc000249830}, 0xc0001d9520, {0xc48940, 0xc00068e230})
        /home/piotr/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.19.0/helper/schema/resource.go:1015 +0x585
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc0004787e0, {0xdbdc60?, 0xc0002496e0?}, 0xc00011a880)
        /home/piotr/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.19.0/helper/schema/grpc_provider.go:613 +0x4a5
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0xc0003beb40, {0xdbdc60?, 0xc000248ab0?}, 0xc0006ca480)
        /home/piotr/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.12.0/tfprotov5/tf5server/server.go:747 +0x423
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0xc570a0?, 0xc0003beb40}, {0xdbdc60, 0xc000248ab0}, 0xc00068e000, 0x0)
        /home/piotr/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.12.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:349 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001ca1e0, {0xdc1b38, 0xc000288d00}, 0xc00034b9e0, 0xc00048f620, 0x127d4d0, 0x0)
        /home/piotr/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1295 +0xb2b
google.golang.org/grpc.(*Server).handleStream(0xc0001ca1e0, {0xdc1b38, 0xc000288d00}, 0xc00034b9e0, 0x0)
        /home/piotr/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1636 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /home/piotr/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:932 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/piotr/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:930 +0x28a

Error: The terraform-provider-nobl9 plugin crashed!
```

**Comment by klaudia-nobl9** [comment]
I've tested it and I haven't found any other bugs other than the problem we've already discussed in regards to removing anomalyConfig from the SLO resource.

For the sake of documentation, I'm adding steps to reproduce mentioned error:
1. Apply the .tf with SLO resource with anomaly_config block with couple of Alert Methods.
2. Remove anomaly_config block and try to reapply the .tf file. 

Actual result: After deleting the entire anomaly_config block, Terraform sees no changes to the configuration and returns the message: `No changes. Your infrastructure matches the configuration.`
Expected result: Terraform should detect configuration changes.

Similar bug was noted during our work on labels: https://github.com/nobl9/terraform-provider-nobl9/pull/45#issuecomment-1220546213

**Comment by klaudia-nobl9** [comment]
Works flawlessly 🌻

File: nobl9/anomaly_config_schema.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+					Description: "Alert Policies attached to SLO",
+					Elem: &schema.Resource{
+						Schema: map[string]*schema.Schema{
+							"alert_method": {
+								Type:             schema.TypeList,
+								Optional:         true,
+								Description:      "Alert methods attached to Anomaly Config",
+								MaxItems:         5,
+								DiffSuppressFunc: diffSuppressAnomalyConfig,
+								Elem: &schema.Resource{
```
[the remark this answers is not in the document]
Comment (nuusk) [reply nobl9/anomaly_config_schema.go:27]: We're not sure how this structure will look like in the future so I think it's better to have it all here for simplicity

File: nobl9/resource_slo.go (hunk 3 above)
Comment (nuusk) [reply nobl9/resource_slo.go:780]: thx, it was a typo :)

File: nobl9/anomaly_config_schema_test.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
+package nobl9
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	n9api "github.com/nobl9/nobl9-go"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
```
Comment (piotrkwarcinski) [inline nobl9/anomaly_config_schema_test.go:8]: please sort these imports before merging

**Review by piotrkwarcinski — DISMISSED** [review]
Tested locally, works now as expected :rocket: , I left one more comment to sort some imports before merging.

**Review by piotrkwarcinski — APPROVED** [review]

**Review by piotrkwarcinski — APPROVED** [review]

**Comment by klaudia-nobl9** [comment]
Did some smoke testing after bumbing nobl9-go, works as expected 🚀

**Pull request merged**